### ISSUE TITLE
Jaeger agent as k8s sidecar on different namespace from observability namespace not working

### ISSUE THREAD

**Issue by yoramda** [jaegertracing/jaeger-client-csharp#201]
Hi,

Jaeger installed as Jaeger operator on observability  namespace 
My app deployment with sidecar annotation:

apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-app-name
  annotations: 
    "sidecar.jaegertracing.io/inject": "true"

Here is the environment setting that was created automatically by the sidecar:
JAEGER_SERVICE_NAME=xxxx-jet-apigateway.xxxx-jet-dev
JAEGER_PROPAGATION=jaeger,b3,w3c

Problem:
On Jaeger UI, Service dropdown shows IP address instead of service name, also can't see service dependency,
I tried to check with all-in-one and production strategy,
Also tried to add from the client library JAEGER_SERVICE_NAME, and also as environment variable : JAEGER_SERVICE_NAME

My client app is DotnetCore c# with openTelemtry nugget : OpenTelemetry.Exporter.Jaeger

Any ideas ?

Thanks in advanced .

**Comment by yoramda** [comment]
tried to use it without agent host & port like this :

services.AddOpenTelemetryTracing((builder) => builder
                            .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService(k8sAppName))
                            .AddAspNetCoreInstrumentation(
                               (options) => options.Filter =
                                   (httpContext) =>
                                   {
                                       // do not collect telemetry metric
                                       return ! httpContext.Request.Path.Value.Equals("/metrics");
                                   }
                               )
                            .AddHttpClientInstrumentation()
                            .AddJaegerExporter(jaegerOptions =>
                            {
                               
                            }));

also tried to add values from enviorment variables like this : 
 
 services.AddOpenTelemetryTracing((builder) => builder
                            .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService(k8sAppName))
                            .AddAspNetCoreInstrumentation(
                               (options) => options.Filter =
                                   (httpContext) =>
                                   {
                                       // do not collect telemetry metric
                                       return ! httpContext.Request.Path.Value.Equals("/metrics");
                                   }
                               )
                            .AddHttpClientInstrumentation()
                            .AddJaegerExporter(jaegerOptions =>
                            {
                               var port = configuration.GetValue<string>($"JAEGER_AGENT_PORT");
                                var host = configuration.GetValue<string>($"JAEGER_AGENT_HOST");
                                jaegerOptions.AgentHost = host;
                                jaegerOptions.AgentPort = string.IsNullOrEmpty(port) ? 6831 : int.Parse(port);
                            })); 

with deployment yaml like this:
env:
	
	- name: JAEGER_AGENT_PORT
	  value: "6831"
	- name: JAEGER_AGENT_HOST
	  valueFrom:
		fieldRef:
		  fieldPath: status.hostIP
	- name: JAEGER_SERVICE_NAME
	  value: my_app_name

did not solve the problem .

**Comment by yoramda** [comment]
I have used the following tracer library 
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/examples/AspNetCore/Startup.cs

**Comment by yoramda** [comment]
whitch library is beter to use for kubernetes support  ?
https://github.com/jaegertracing/jaeger-client-csharp   
OR
https://github.com/open-telemetry/opentelemetry-dotnet

**Comment by Falco20019** [comment]
The open telemetry exporter has it‘s own implementation and is, AFAIK, not using this library in any way. So if you plan on using that library, you need to ask at https://github.com/open-telemetry/opentelemetry-dotnet

jaeger-client-csharp should work fine with k8s and allows instantiation from environment. Examples on how to create your tracer from environment can be found in the main README of this repository.

**Comment by yoramda** [comment]
Hi,

I have refactor my code to use the recommended library 👍 https://github.com/jaegertracing/jaeger-client-csharp,
now the agent is not sending the traces to the collector at all.

here is my client code 
`// Adds the Jaeger Tracer.
            services.AddSingleton<ITracer>(serviceProvider =>
            {
                string serviceName = serviceProvider.GetRequiredService<IWebHostEnvironment>().ApplicationName;

                var loggerFactory = serviceProvider.GetRequiredService<ILoggerFactory>();
                var sampler = new ConstSampler(sample: true);
                var reporter = new RemoteReporter.Builder()
                 .WithLoggerFactory(loggerFactory)
                 .WithSender(new UdpSender(configuration.GetValue<string>("JAEGER_AGENT_HOST"), 6831, 0))
                 .Build();
                var tracer = new Tracer.Builder(serviceName)
                 .WithLoggerFactory(loggerFactory)
                 .WithSampler(sampler)
                 .WithReporter(reporter)
                 .Build();
                if (!GlobalTracer.IsRegistered())
                {
                    GlobalTracer.Register(tracer);
                }
                return tracer;

            });`

try to setup the JAEGER_AGENT_HOST env variable to :
`env:
- name: JAEGER_AGENT_HOST
  valueFrom:
	fieldRef:
	  fieldPath: status.hostIP

did not worked, also tried to set it up to localhost / 127.0.0.1 did not work .

I am using jaeger operator on k8s vanila with this setup:

`apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: simplest
spec:
  strategy: production
  storage:
    type: elasticsearch
    options:
      es:
        # Note: This assumes elasticsearch is running in the "default" namespace.
        server-urls: http://elastic:31221 `

any ideas ?

**Comment by Falco20019** [comment]
I assume this code was taken from https://github.com/opentracing-contrib/csharp-netcore/blob/master/samples/netcoreapp3.1/Shared/JaegerServiceCollectionExtensions.cs ? When using ASP.NET Core, you might want to use opentracing-contrib/csharp-netcore which helps with binding all the calls. The code you wrote only instantiates the tracer but doesn’t do anything by itself. 

https://github.com/opentracing-contrib/csharp-netcore/blob/e5b0c8a6fc1321ff87f558b69c5ad1a95912acc3/samples/netcoreapp3.1/OrdersApi/Program.cs#L27-L37

You also have to setup the other half using the tracer. Maybe that’s what you are missing? The code that you posted is only the AddJaeger part which is independent of if you use ASP, Core oder Framework.

**Comment by yoramda** [comment]
Hi,
My goal is to setup jaeger operator for k8s with elastic storage and production strategy.
I am tiring to use jaeger-agent as a sidecar, by adding jaeger-agent annotation to my app deployment, is it good or bad practice ? is it better to report directly to the collector  ?
jaeger-agent with udpSender using UDP, is it better from performance perspective ?

I can see that on 
https://github.com/opentracing-contrib/csharp-netcore/blob/e5b0c8a6fc1321ff87f558b69c5ad1a95912acc3/samples/netcoreapp3.1/Shared/JaegerServiceCollectionExtensions.cs#L18
the code uses this line of code 
private static readonly Uri _jaegerUri = new Uri("http://localhost:14268/api/traces");

I will try to use the example you have provided and update on the results .

Thanks in advanced.

**Comment by Falco20019** [comment]
Usually it‘s the default case to have jaeger-agent as endpoint from your application as side car since UDP has a lot less overhead than HTTP over TCP. So the side car collects some traces and then aggregates them to reduce the HTTP calls to the collector. This should also increase the throughput and unblock your application from waiting for HTTP calls to safely return.

The code you mentioned writes to the collector directly. That’s fine for examples but not for high throughput.

Setup is a lot easier when using environment variables. You just have to set the host and port variables and UDP will automatically be selected when instantiated like written above.

**Comment by yoramda** [comment]
Hi,

I have tried to call the "http://localhost:14268/api/traces", but that didn't help, still no traces are reported at all.
i am starting to think that i may have a network access problem between my app deployment namespace and the observability namespace, i tried to setup the the role binding as it says at the Jaeger documentation 

https://www.jaegertracing.io/docs/1.21/operator/ :

"
When not using the cluster-wide resources (ClusterRole and ClusterRoleBinding), set the WATCH_NAMESPACE to the comma-separated list of namespaces that the Jaeger Operator should watch for events related to Jaeger resources. **It is possible to have the Jaeger Operator running in a given namespace (like, observability) and manage Jaeger resources in another (like, myproject). For that, use a RoleBinding like the following for each namespace the operator should watch for resources:**

kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: jaeger-operator-in-myproject
  namespace: myproject
subjects:
- kind: ServiceAccount
  name: jaeger-operator
  namespace: observability
roleRef:
  kind: Role
  name: jaeger-operator
  apiGroup: rbac.authorization.k8s.io
"

any ideas ?
Thanks in advanced .

**Comment by Falco20019** [comment]
Sorry, but I‘m not too deep into k8s myself. So I‘m not a big help here. But maybe someone else is able to help you. @yurishkuro do you know some more?

**Comment by yoramda** [comment]
Here is the jaeger-agent (sidecar) container arguments

--agent.tags=cluster=undefined,container.name=xxx-customermanagement-api,deployment.name=xxx-customermanagement-api,host.ip=${HOST_IP:},pod.name=${POD_NAME:},pod.namespace=xxx-dev5
--reporter.grpc.host-port=dns:///simplest-collector-headless.observability.svc:14250

**Comment by yoramda** [comment]
Hi,

The jager-agent as a sidecar is not working, i will open another issue to the jaeger operator gitub.
When i am bypassing the local sidecar agent and call the collector directly with httpSender

`IReporter reporter = new RemoteReporter.Builder()
                //.WithLoggerFactory(loggerFactory)
                   .WithSender(new HttpSender.Builder(new Uri("http://simplest-collector.observability:14268/api/traces").ToString()).Build())
                   .Build();`

It started to work 👍 but, i am using w3c that generates traceID in 32 bit format, for example : 
`"traceID": "57614cf030bac742b004904892a8f448", `
i am using this traceID in my logs across all microservices .
`Activity.DefaultIdFormat = ActivityIdFormat.W3C;`

the jaeger reporting the trace in shorter format, for example looking at the TraceJson 
`traceID": "090693e81d761189"`

what is the correct format to fix it   ?

Thanks in advanced .

**Comment by yurishkuro** [comment]
Jaeger SDKs support both short (64bit) and long (128bit) IDs. Short IDs are on by default, you need to enable 128bit IDs.

**Comment by Falco20019** [comment]
Using `Configuration.FromEnv()`, you would need to set the following variables:

```
JAEGER_SERVICE_NAME = xyz 
JAEGER_ENDPOINT = http://simplest-collector.observability:14268/api/traces
JAEGER_TRACEID_128BIT = true
```

**Comment by yoramda** [comment]
Hi,

JAEGER_TRACEID_128BIT = true solved the problem and now it is  generating 128bit trace id format, but it looks like the w3c traceId on my console log is different that the one created by Jaeger Trace object.

Any Ideas ?
Thanks in advanced.

**Comment by Falco20019** [comment]
Sorry, I missed that you use w3c traces. There is an open issue to implement the w3c headers: https://github.com/jaegertracing/jaeger-client-csharp/issues/90

In the README for `JAEGER_PROPAGATION` you can see, that only `jaeger` and `b3` are supported right now. If you are forced to use `w3c`, I could check what's missing.

**Comment by yoramda** [comment]
Thanks,

to reproduce the problem , i have modify the program.cs on the CustomerApi as follow and add the ConfigureLogging.
is it possible to print to the console log the traceId & spanId & parentId, for example : 

[22-02-2021 10:03:43]info: Microsoft.AspNetCore.Mvc.StatusCodeResult[1]
      => ConnectionId:0HM6N11EEK2UU => RequestPath:/customers/100 RequestId:0HM6N11EEK2UU:00000001, SpanId:cfd85f432eda174f, TraceId:afa5e110ff80944a8c6b4434c20cfbbe, ParentId:0000000000000000 => Samples.CustomersApi.Controllers.Customer


`
public class Program
    {
        public static void Main(string[] args)
        {
            **Activity.DefaultIdFormat = ActivityIdFormat.W3C;**
            CreateHostBuilder(args).Build().Run();
        }

        public static IHostBuilder CreateHostBuilder(string[] args)
        {
            return Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder
                        .UseStartup<Startup>()
                        .UseUrls(Constants.CustomersUrl);
                })
               .ConfigureLogging(logging =>
               {
                   logging.SetMinimumLevel(Microsoft.Extensions.Logging.LogLevel.Warning);
                   logging.AddConsole((options) => { options.IncludeScopes = true; options.TimestampFormat = "[dd-MM-yyyy HH:mm:ss]"; });
               })
                .ConfigureServices(services =>
                {
                    // Registers and starts Jaeger (see Shared.JaegerServiceCollectionExtensions)
                    services.AddJaeger();

                    // Enables OpenTracing instrumentation for ASP.NET Core, CoreFx, EF Core
                    services.AddOpenTracing(builder =>
                    {
                        builder.ConfigureAspNetCore(options =>
                        {
                            // We don't need any tracing data for our health endpoint.
                            options.Hosting.IgnorePatterns.Add(ctx => ctx.Request.Path == "/health");
                        });

                        builder.ConfigureEntityFrameworkCore(options =>
                        {
                            // This is an example for how certain EF Core commands can be ignored.
                            // As en example, we're ignoring the "PRAGMA foreign_keys=ON;" commands that are executed by Sqlite.
                            // Remove this code to see those statements.
                            options.IgnorePatterns.Add(cmd => cmd.Command.CommandText.StartsWith("PRAGMA"));
                        });
                    });

                });
        }
    }
`

**Comment by Falco20019** [comment]
The main problem with w3c is, that the header names are different. So we need to implement the propagation that knows where to read the IDs from. In your case, even when you get the request from a w3c trace, it will just create a new one instead of appending that one, since it doesn't know the header.

We already started implementing it, but never finished it since there was no need yet: https://github.com/jaegertracing/jaeger-client-csharp/compare/w3c-trace-headers

So if you can't use `jaeger` or `b3` format, I would try to rebase that effort and finalize it for you to test.

**Comment by yoramda** [comment]
Thanks,

How can i switch to jaeger or b3 format ?

**Comment by Falco20019** [comment]
`JAEGER_REPORTER_LOG_SPANS = true` for environment. Or the `LoggingReporter` like you did in a `CompositeReporter` together with the `RemoterReporter` using a `HttpSender` (a lot more complex).

**Comment by yoramda** [comment]
Thanks a lot, It's working great !!
Regarding the W3C propagation, I will  be happy to assist on testing it.

**Comment by yoramda** [comment]
Hi,

JAEGER_REPORTER_LOG_SPANS = true printing the :
[23-02-2021 15:16:53]info: Jaeger.Reporters.LoggingReporter[0]
      Span reported: c94e123685f801601cbb905515f40081:1cbb905515f40081:0000000000000000:1 - HTTP GET
for Jaeger.* classes only, but not for my custom class MyClassTest .

Here is my appsettings.json 
"Logging": {
    "LogLevel": {
      "Default": "Warning",
      "Microsoft": "Warning",
      "Microsoft.Hosting.Lifetime": "Warning",
      "MyClassTest.*": "Trace",
      "Jaeger.*": "Trace",
      "OpenTracing.Contrib.NetCore.*": "Trace"
    }
  },

**Comment by Falco20019** [comment]
What is your custom class doing? `JAEGER_REPORTER_LOG_SPANS` only enables logging problems and the finishing of spans. Your custom class needs to use `ILoggerFactory` as usual by itself.

```csharp
var logger = loggerFactory.CreateLogger<MyClassTest>();
logger.LogInfo(„HelloWorld“);
```

What namespace is your class in? Your configuration needs to match the qualified name with namespace and not just the class name. But that’s default Microsoft Logging and not related to Jaeger itself as we are also just using the default framework.

**Comment by yoramda** [comment]
Thanks,

I have implemented custom console provider for that .

**Comment by yoramda** [comment]
Hi,

First, this a great library, thanks for the Good work !

Logs (when opening the accordion) on the Jaeger-UI is a bit strange .

For example:
`Microsoft.AspNetCore.Routing.EndpointMatched is not exists in the console log, but exists in the Logs UI `

Another Example, exsists on the console log, but not showing on the Logs UI,  sql is not logged 
`[2021-02-24 04:15:03]-[Information] Microsoft.EntityFrameworkCore.Infrastructure TraceId=6fc2f145388e62044650951bb595ae6c, SpanId=ac5e272a3c519d68
-----> Entity Framework Core 3.1.7 initialized 'JetJobContext' using provider 'Npgsql.EntityFrameworkCore.PostgreSQL' with options: NoTracking SensitiveDataLoggingEnabled DetailedErrorsEnabled 
[2021-02-24 04:15:03]-[Information] Microsoft.EntityFrameworkCore.Database.Command TraceId=6fc2f145388e62044650951bb595ae6c, SpanId=ac5e272a3c519d68
-----> Executed DbCommand (1ms) [Parameters=[@__activity_JobId_0='xxx-6d88637497728576059036637497728877308946'], CommandType='Text', CommandTimeout='30']
SELECT job_id
FROM test AS j
WHERE job_id = @__activity_JobId_0
LIMIT 2`

Any way to configure ?

Thanks .

**Comment by Falco20019** [comment]
There are two logging contexts. One from the logger framework (ILoggerFactory) and a second one from the Jaeger Tracing framework. The first ones are not automatically logged into the second or vice versa and also not linked together.

There are ways to connect both, like using Serilog as logger framework (compatible to Microsoft.Extensions.Logging) and using an Enricher to add the current traceID and spanID to the log context.

There is a difference between logging and tracing. Logging is mostly about a time sequence of entries only differentiated by their level and also sometimes enhanced with additional data like key value pairs. It does not really have any information about how your application is linked together or also not a lot information about the environment.

Tracing is focusing on how everything is connected. Internally and externally if the external system allows to follow the trace. The logs in there are not related to any logging framework other than OpenTracing itself. So they are just pieces of additional information in relation to a trace (or to be more specific: of a span).

The only thing that Jaeger will use the regular logging framework for is for posting success messages about when a span was finished or if something goes wrong in the setup.

I hope this helps a bit to clarify things.

**Comment by yoramda** [comment]
Thanks for your explanation, the tutorial was very useful !

I have implemented custom console logger, and injected the TraceId, SpanId, ParentId .
Also inside the custom logger(logger framework (ILoggerFactory)) I am writing the logs to the jaeger tracer span.Log(), do you see any problem with this approach ? by the way it is working fine .

I have already implemented  Prometheus and Grafana  for metrics and alerts, not use it for logs.
i will take a look on Loki later .

Thanks once again .

**Comment by Falco20019** [comment]
It‘s fine to log to the traces of you don’t have a separate logging endpoint. If you have a separate one, you would try to avoid the duplication and instead link the traces to the logger by annotating the logs with the traceID + spanID.

I will close the issue since your problem is solved. You can still add messages to it if you feel a need for it or create a new one if it‘s not fitting this issue anymore.
